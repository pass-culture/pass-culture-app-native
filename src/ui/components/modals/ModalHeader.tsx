import { t } from '@lingui/macro'
import React, { FunctionComponent } from 'react'
import { LayoutChangeEvent } from 'react-native'
import styled from 'styled-components/native'

import { accessibilityAndTestId } from 'libs/accessibilityAndTestId'
import { getSpacing, Typo } from 'ui/theme'
import { getHeadingAttrs } from 'ui/theme/typography'

import { ModalIconProps } from './types'

type ModalHeaderProps = {
  title: string
  boldTitle?: boolean
  numberOfLines?: number
  onLayout?: (event: LayoutChangeEvent) => void
  accessibilityDescribedBy?: string
} & ModalIconProps

export const ModalHeader: FunctionComponent<ModalHeaderProps> = ({
  title,
  leftIcon,
  leftIconAccessibilityLabel = t`Revenir en arriÃ¨re`,
  onLeftIconPress,
  rightIcon,
  rightIconAccessibilityLabel = t`Fermer la modale`,
  onRightIconPress,
  boldTitle = false,
  numberOfLines = 2,
  onLayout,
  accessibilityDescribedBy,
}) => {
  const TitleComponent = boldTitle ? BoldTitle : Title
  const LeftIcon =
    !!leftIcon &&
    styled(leftIcon).attrs(({ theme }) => ({
      size: theme.icons.sizes.smaller,
    }))``
  const RightIcon =
    !!rightIcon &&
    styled(rightIcon).attrs(({ theme }) => ({
      size: theme.icons.sizes.smaller,
    }))``

  return (
    <Container onLayout={onLayout} testID="modalHeader">
      <LeftHeaderActionContainer>
        {!!LeftIcon && (
          <HeaderAction
            onPress={onLeftIconPress}
            {...accessibilityAndTestId(leftIconAccessibilityLabel)}>
            <LeftIcon testID="leftIcon" />
          </HeaderAction>
        )}
      </LeftHeaderActionContainer>
      <TitleContainer>
        <TitleComponent numberOfLines={numberOfLines} nativeID={accessibilityDescribedBy}>
          {title}
        </TitleComponent>
      </TitleContainer>
      <RightHeaderActionContainer>
        {!!RightIcon && (
          <HeaderAction
            onPress={onRightIconPress}
            {...accessibilityAndTestId(rightIconAccessibilityLabel)}>
            <RightIcon testID="rightIcon" />
          </HeaderAction>
        )}
      </RightHeaderActionContainer>
    </Container>
  )
}

const Container = styled.View({
  display: 'flex',
  width: '100%',
  flexDirection: 'row',
  justifyContent: 'center',
})

const TitleContainer = styled.View(({ theme }) => ({
  justifyContent: 'center',
  paddingRight: getSpacing(3),
  paddingLeft: getSpacing(3),
  flex: 0.8,
  zIndex: theme.zIndex.modalHeader,
}))

const LeftHeaderActionContainer = styled.View({
  flexDirection: 'row',
  flex: 0.1,
  alignItems: 'flex-start',
  justifyContent: 'flex-start',
})

const RightHeaderActionContainer = styled.View({
  flexDirection: 'row',
  flex: 0.1,
  alignItems: 'flex-start',
  justifyContent: 'flex-end',
})

const HeaderAction = styled.TouchableOpacity({
  padding: getSpacing(1),
})

const Title = styled(Typo.Title4).attrs(() => getHeadingAttrs(1))({
  textAlign: 'center',
})
const BoldTitle = styled(Typo.Title3).attrs(() => getHeadingAttrs(1))({
  textAlign: 'center',
})
