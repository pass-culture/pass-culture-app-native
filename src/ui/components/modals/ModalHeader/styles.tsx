import { StyleSheet, TextStyle, ViewStyle } from 'react-native'

export interface ModalHeaderStyleClasses {
  container?: ViewStyle
  leftSide?: ViewStyle
  title?: TextStyle
  rightSide?: ViewStyle
}

export const ModalHeaderStyles = StyleSheet.create<ModalHeaderStyleClasses>({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  leftSide: {
    flex: 0.2,
    alignItems: 'center',
  },
  title: {
    flex: 0.6,
    textAlign: 'center',
  },
  rightSide: {
    flex: 0.2,
    alignItems: 'center',
  },
})
