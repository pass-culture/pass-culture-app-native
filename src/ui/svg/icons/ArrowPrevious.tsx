import * as React from 'react'
import Svg, { Path } from 'react-native-svg'
import styled from 'styled-components/native'

import { IconInterface } from './types'

const ArrowPreviousSvg: React.FunctionComponent<IconInterface> = ({ size, color, testID }) => (
  <Svg width={size} height={size} viewBox="0 0 48 49" testID={testID} aria-hidden>
    <Path
      fill={color}
      fillRule="evenodd"
      clipRule="evenodd"
      d="M31.6748 44.2116C32.07 43.8227 32.075 43.1873 31.6858 42.7924L14.3334 25.28C13.8973 24.8373 13.9038 24.1152 14.3289 23.6893L14.3302 23.688L31.7078 6.21083C32.0988 5.81768 32.0967 5.18229 31.7033 4.79165C31.3098 4.40101 30.674 4.40305 30.2831 4.79621L12.9068 22.272C11.6954 23.4863 11.7016 25.4695 12.9022 26.6881L30.2546 44.2005C30.6438 44.5955 31.2796 44.6004 31.6748 44.2116Z"
    />
  </Svg>
)

export const ArrowPrevious = styled(ArrowPreviousSvg).attrs(({ color, size, theme }) => ({
  color: color ?? theme.colors.black,
  size: size ?? theme.icons.sizes.standard,
}))``
