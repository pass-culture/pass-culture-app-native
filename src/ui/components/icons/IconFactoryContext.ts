import { createContext } from 'react'

import { IconFactory } from 'ui/components/icons/iconFactory'

export const IconFactoryContext = createContext<IconFactory>({
  getIcon: () => {
    throw new Error('No <IconFactoryProvider> detected')
  },
})
