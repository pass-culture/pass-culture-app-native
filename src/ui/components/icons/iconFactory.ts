import { ArrowLeftNew } from 'ui/svg/icons/ArrowLeftNew'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { ArrowRightNew } from 'ui/svg/icons/ArrowRightNew'
import { Share } from 'ui/svg/icons/BicolorShare'
import { Favorite } from 'ui/svg/icons/Favorite'
import { FavoriteFilled } from 'ui/svg/icons/FavoriteFilled'
import { ThumbDown } from 'ui/svg/icons/ThumbDown'
import { ThumbDownFilled } from 'ui/svg/icons/ThumbDownFilled'
import { ThumbUp } from 'ui/svg/icons/ThumbUp'
import { ThumbUpFilled } from 'ui/svg/icons/ThumbUpFilled'
import { AccessibleIcon } from 'ui/svg/icons/types'

export type IconNames =
  | 'back'
  | 'share'
  | 'favorite'
  | 'favorite-filled'
  | 'next'
  | 'previous'
  | 'like'
  | 'like-filled'
  | 'dislike'
  | 'dislike-filled'

export type IconFactory = { getIcon(iconName?: IconNames): React.FC<AccessibleIcon> }

const iconMapping: Record<IconNames, React.FC<AccessibleIcon>> = {
  back: ArrowPrevious,
  next: ArrowRightNew,
  previous: ArrowLeftNew,
  share: Share,
  'favorite-filled': FavoriteFilled,
  favorite: Favorite,
  like: ThumbUp,
  'like-filled': ThumbUpFilled,
  dislike: ThumbDown,
  'dislike-filled': ThumbDownFilled,
}

export const iconFactory: IconFactory = {
  getIcon(name: IconNames) {
    return iconMapping[name]
  },
}
