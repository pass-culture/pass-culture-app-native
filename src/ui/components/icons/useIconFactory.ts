import { useContext } from 'react'

import { IconFactoryContext } from 'ui/components/icons/IconFactoryContext'

export const useIconFactory = () => useContext(IconFactoryContext)
