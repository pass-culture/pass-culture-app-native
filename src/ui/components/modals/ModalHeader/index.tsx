import React, { FunctionComponent } from 'react'
import { Text, TouchableOpacity, View } from 'react-native'

import { IconInterface } from 'ui/icons/types'
import { Typo } from 'ui/theme'

import { ModalHeaderStyleClasses, ModalHeaderStyles } from './styles'

interface Props {
  title: string
  leftIcon?: FunctionComponent<IconInterface>
  onLeftIconPress?: () => void
  rightIcon?: FunctionComponent<IconInterface>
  onRightIconPress?: () => void
  classes?: ModalHeaderStyleClasses
}

const ModalHeader: FunctionComponent<Props> = ({
  classes,
  title,
  leftIcon: LeftIcon,
  onLeftIconPress,
  rightIcon: RightIcon,
  onRightIconPress,
}) => {
  return (
    <View style={[ModalHeaderStyles.container, classes?.container]}>
      <TouchableOpacity
        style={ModalHeaderStyles.leftSide}
        onPress={onLeftIconPress}
        testID="leftIconButton">
        {LeftIcon && <LeftIcon size={32} testID="leftIcon" />}
      </TouchableOpacity>
      <Text style={[ModalHeaderStyles.title, classes?.title]}>
        <Typo.Title4 numberOfLines={2}>{title}</Typo.Title4>
      </Text>
      <TouchableOpacity
        style={ModalHeaderStyles.rightSide}
        onPress={onRightIconPress}
        testID="rightIconButton">
        {RightIcon && <RightIcon size={32} testID="rightIcon" />}
      </TouchableOpacity>
    </View>
  )
}

export default ModalHeader
