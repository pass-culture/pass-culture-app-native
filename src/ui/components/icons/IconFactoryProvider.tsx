import React, { PropsWithChildren } from 'react'

import { iconFactory } from 'ui/components/icons/iconFactory'
import { IconFactoryContext } from 'ui/components/icons/IconFactoryContext'

export const IconFactoryProvider = ({ children }: PropsWithChildren) => (
  <IconFactoryContext.Provider value={iconFactory}>{children}</IconFactoryContext.Provider>
)
