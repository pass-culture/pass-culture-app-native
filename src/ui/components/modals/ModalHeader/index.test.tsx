import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { ArrowPrevious } from 'ui/icons/ArrowPrevious'
import { Close } from 'ui/icons/Close'

import ModalHeader from '.'

jest.mock('libs/environment', () => ({
  env: {
    FEATURE_FLAG_CODE_PUSH: true,
  },
}))

describe('ModalHeader component', () => {
  it('should render correctly', () => {
    const home = render(<ModalHeader title="Testing modal header rendering" />)
    expect(home).toMatchSnapshot()
  })
  describe('left icon', () => {
    it('should be hidden when the icon is not provided', () => {
      const { getByTestId } = render(<ModalHeader title="Testing modal header rendering" />)
      try {
        const leftIcon = getByTestId('leftIcon')
        expect(leftIcon).toBeNull()
      } catch (error) {
        expect(error).toBeDefined()
      }
    })
    it('should be visible when the icon is provided', async () => {
      const onLeftPress = jest.fn()
      const { getByTestId, findByTestId } = render(
        <ModalHeader
          title="Testing modal header rendering"
          leftIcon={ArrowPrevious}
          onLeftIconPress={onLeftPress}
        />
      )

      getByTestId('leftIcon')

      const leftIconButton = await findByTestId('leftIconButton')
      await fireEvent.press(leftIconButton)
      expect(onLeftPress).toBeCalledTimes(1)
    })
  })
  describe('right icon', () => {
    it('should be hidden when the icon is not provided', () => {
      const { getByTestId } = render(<ModalHeader title="Testing modal header rendering" />)
      try {
        const rightIcon = getByTestId('rightIcon')
        expect(rightIcon).toBeNull()
      } catch (error) {
        expect(error).toBeDefined()
      }
    })
    it('should be visible when the icon is provided', async () => {
      const onRightPress = jest.fn()
      const { getByTestId, findByTestId } = render(
        <ModalHeader
          title="Testing modal header rendering"
          rightIcon={Close}
          onRightIconPress={onRightPress}
        />
      )

      getByTestId('rightIcon')

      const rightIconButton = await findByTestId('rightIconButton')
      await fireEvent.press(rightIconButton)
      expect(onRightPress).toBeCalledTimes(1)
    })
  })
})
