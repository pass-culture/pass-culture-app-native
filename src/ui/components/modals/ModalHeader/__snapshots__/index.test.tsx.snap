// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalHeader component should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 0.2,
        "opacity": 1,
      }
    }
    testID="leftIconButton"
  />
  <Text
    style={
      Array [
        Object {
          "flex": 0.6,
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    <Text
      numberOfLines={2}
      style={
        Array [
          Object {
            "color": "#151515",
            "fontFamily": "Montserrat-Medium",
            "fontSize": 18,
            "fontWeight": "500",
            "lineHeight": 22,
          },
        ]
      }
    >
      Testing modal header rendering
    </Text>
  </Text>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 0.2,
        "opacity": 1,
      }
    }
    testID="rightIconButton"
  />
</View>
`;
